keycloak.d.ts

```diff
@@ -0,0 +1,348 @@
+    scope?: string
+    store?: any
+    cookies?: boolean
+  }
+
+  interface GrantProperties {
+    access_token?: string
+    refresh_token?: string
+    id_token?: string
+    expires_in?: string
+    token_type?: string
+  }
+
+  interface Token {
+    isExpired(): boolean
+    hasRole(): boolean
+    hasApplicationRole(): boolean
+    hasRealmRole(): boolean
+  }
+
+  interface GrantManager {
+    /**
+     * Use the direct grant API to obtain a grant from Keycloak.
+     *
+     * The direct grant API must be enabled for the configured realm
+     * for this method to work. This function ostensibly provides a
+     * non-interactive, programatic way to login to a Keycloak realm.
+     *
+     * @param {String} username The username.
+     * @param {String} password The cleartext password.
+     */
+    obtainDirectly(username: string, password: string): Promise<Grant>
+
+    /**
+     * Obtain a grant from a previous interactive login which results in a code.
+     *
+     * This is typically used by servers which receive the code through a
+     * redirect_uri when sending a user to Keycloak for an interactive login.
+     *
+     * An optional session ID and host may be provided if there is desire for
+     * Keycloak to be aware of this information.  They may be used by Keycloak
+     * when session invalidation is triggered from the Keycloak console itself
+     * during its postbacks to `/k_logout` on the server.
+     *
+     * @param {String} code The code from a successful login redirected from Keycloak.
+     * @param {String} sessionId Optional opaque session-id.
+     * @param {String} sessionHost Optional session host for targetted Keycloak console post-backs.
+     */
+    obtainFromCode(code: string, sessionid: string, sessionHost: string): Promise<Grant>
+
+
+    // obtainFromClientCredentials
+
+    /**
+     * Ensure that a grant is *fresh*, refreshing if required & possible.
+     *
+     * If the access_token is not expired, the grant is left untouched.
+     *
+     * If the access_token is expired, and a refresh_token is available,
+     * the grant is refreshed, in place (no new object is created),
+     * and returned.
+     *
+     * If the access_token is expired and no refresh_token is available,
+     * an error is provided.
+     *
+     * @param {Grant} grant The grant object to ensure freshness of
+     */
+    ensureFreshness (grant: Grant): Promise<Grant>
+
+    /**
+     * Perform live validation of an `access_token` against the Keycloak server.
+     *
+     * @param {Token|String} token The token to validate.
+     * @param {Function} callback Callback function if not using promises.
+     *
+     * @return {boolean} `false` if the token is invalid, or the same token if valid.
+     */
+    validateAccessToken(token: string): Promise<boolean>
+
+    /**
+     * Returns a user info JSON Object
+     * @param {String} token
+     */
+    userInfo(token: string): Promise<Object>
```

Comment:
Yeah, Object was a placeholder. I've found [this](https://www.keycloak.org/docs/3.0/securing_apps/topics/oidc/oidc-generic.html) that describes it as "standard claims". @abstractj, do you know if this object is defined someplace? Does it contain known properties, or are some dynamic?

I think it could be represented as something like this so application specific keys can be defined:

```ts
userInfo<T extends StandardClaims>(token: string): Promise<T>
```

Reply:
```ts
userInfo<T extends Token|string, C extends StandardClaims>(token: T): Promise<C>
```